src/main/java/org/apache/maven/plugins/checkstyle/Violation.java

```diff
@@ -0,0 +1,229 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.util.Objects;
+import java.util.StringJoiner;
+
+/**
+ * Holds data about a single violation and represents the violation itself.
+ */
+public class Violation
+{
+
+  /**
+   * Indicates that a column is not set.
+   */
+  protected static final String NO_COLUMN = "-1";
+
+  /** The source file name relative to the project's root. */
+  private final String source;
+
+  /** File is the absolute file name to the checked file. */
+  private final String file;
+
+  private final String line;
+
+  private String column = NO_COLUMN;
+
+  private final String severity;
+
+  private final String message;
+
+  private final String ruleName;
+
+  private final String category;
+
+  // Leaving out column, because there is no CHECKSTYLE:OFF support.
+
+  /**
+   * Creates a violation instance without a column set.
+   *
+   * @param source
+   *     the source file name, relative to the project's root.
+   * @param file
+   *     the absolute file name in which the violation occurred.
+   * @param line
+   *     the line in the file on which the violation occurred.
+   * @param severity
+   *     the severity of the violation.
+   * @param message
+   *     the message from checkstyle for this violation.
+   * @param ruleName
+   *     the rule name from which this violation was created.
+   * @param category
+   *     the category of the checkstyle violation.
+   */
+  public Violation( String source,
+                    String file,
+                    String line,
+                    String severity,
+                    String message,
+                    String ruleName,
+                    String category )
+  {
+    this.source = Objects.requireNonNull( source );
+    this.file = file;
+    this.line = line;
+    this.severity = Objects.requireNonNull( severity );
+    this.message = Objects.requireNonNull( message );
+    this.ruleName = Objects.requireNonNull( ruleName );
+    this.category = Objects.requireNonNull( category );
+  }
+
+  /**
+   * Returns the source file name relative to the project's root.
+   *
+   * @return the source file name relative to the project's root.
+   */
+  protected String getSource( )
+  {
+    return source;
+  }
+
+  /**
+   * Returns the absolute file name to the checked file.
+   *
+   * @return the absolute file name to the checked file.
+   */
+  protected String getFile( )
+  {
+    return file;
+  }
+
+  /**
+   * Returns the line of in the checked file on which the violation occurred.
+   *
+   * @return the line of in the checked file on which the violation occurred.
+   */
+  protected String getLine( )
+  {
+    return line;
+  }
+
+  /**
+   * Returns the column in which the violation occurred, if available.
+   *
+   * @return the column in which the violation occurred, if available. Otherwise returns {@link #NO_COLUMN}.
+   */
+  protected String getColumn( )
```

Comment:
why is this a string instead of an int?

Reply:
Because  either @rfscholte  or @eolivelli asked me not to use Integer.parseInt() on the XML input. I think this should stay a string for now, because we are not using any comparison against this field at the moment.

Reply:
Because this is published API, we can't easily change it after we publish it once. If it should be a string, fine; but we need to decide the right type now.

Reply:
It's only protected, and it can be nullable. So it's either `OptionalInt` or `@Nullable Integer` or `int` with a null value pattern -- or `@Nullable String` as right now. Each method has its advantages and disadvantages.

I originally had it as int, because a column must always be a number until somone complained about the NumberFormatException (in that case I had a catch block with `null`, but that did not seem enough). Your choice.

Reply:
In the case of NumberFormatException I think -1 == NO_COLUMN works.

Reply:
Why does it need to be protected? Would package private work instead? That would be easier to change later.

Reply:
Still, why use parseInt in the first place? Where's the gain? And as requested thrice now, we'd rather like to change the constructor to have a truly immutable object. Can you help with this?